edgekv/src/datastore.rs

```
@@ -0,0 +1,835 @@
+}
+
+pub fn fetch_double_buffer_file(buffer_dir: &BTreeMap<i64, PathBuf>) -> Result<HashMap<Vec<u8>, DataEntry>> {
+    let mut double_buffer = HashMap::new();
+    for (_, file_path) in buffer_dir.iter() {
+        let double_buffer_file = File::open(file_path)?;
+        let mut rdr = BufReader::new(double_buffer_file);
+        while let Ok(data_entry) = DataEntry::decode(&mut rdr) {
+            if !data_entry.check_crc() {
+                return Err(EdgeKVError::CorruptData);
+            }
+            double_buffer.insert(data_entry.key(), data_entry);
+        }
+    }
+    Ok(double_buffer)
+}
+
+impl DataStore {
+    pub fn open<P: AsRef<Path>>(dir: P) -> Result<Self> {
+        let lock_file = get_lock_file(dir.as_ref())?;
+        let active_file_pair = create_new_file_pair(dir.as_ref())?;
+        let (files_dir, buffer_files) = fetch_file_pairs(dir.as_ref())?;
+        let double_buffer = fetch_double_buffer_file(&buffer_files)?;
+        let keys_dir = KeysDir::new(&files_dir)?;
+        let index_dir = IndexDir::new(files_dir)?;
+        let mut instance = Self {
+            lock_file,
+            dir: dir.as_ref().to_path_buf(),
+            active_file: RwLock::new(ActiveFilePair::from(active_file_pair)?),
+            keys_dir,
+            index_dir,
+            buffer: RwLock::new(Default::default()),
+            double_buffer,
+            buffer_size: RwLock::new(0)
+        };
+        instance.lock()?;
+        Ok(instance)
+    }
+
+    fn lock(&mut self) -> Result<()> {
+        self.lock_file
+            .lock_exclusive()
+            .map_err(|_| EdgeKVError::LockFailed(String::from(self.dir.to_string_lossy())))?;
+        Ok(())
+    }
+
+    pub fn keys_dir(&self) -> &KeysDir {
+        &self.keys_dir
+    }
+
+    pub fn buffer_size(&self) -> usize {
+        let mut buffer = match self
+            .buffer
+            .read() {
+            Ok(buff) => {
+                buff
+            }
+            Err(_) => {
```

Another silently ignored error